lib/api/authentication/grader.py

```diff
@@ -46,9 +46,18 @@ def authenticate(self, request: Request) -> Optional[Tuple[GraderUser, Payload]]
     def add_token_permissions(self, token: str, permissions: ObjectPermissions, payload: Payload):
         if token[0] == "s":
             submission = self.authenticate_submission_token(token[1:])
+            payload.permissions.submissions.add(Permission.READ, id=submission.id)
+            permissions.submissions.add(Permission.READ, submission)
             payload.permissions.submissions.add(Permission.WRITE, id=submission.id)
             permissions.submissions.add(Permission.WRITE, submission)
             exercise = submission.exercise
+
+            # MOOC-jutut needs this but it doesn't seem correct to give exercise read
+            # permissions here. It might not _actually_ even need this but the way
+            # the permissions system works, accessing the submission also requires
+            # access to the exercise.
+            payload.permissions.exercises.add(Permission.READ, id=exercise.id)
```

Comment:
Now that both if branches set exercise READ permission, these two lines could moved from the if-elif branches to the end with the course permissions.